descartes_light/core/include/descartes_light/types.h

```diff
@@ -3,28 +3,188 @@
 #include <descartes_light/descartes_macros.h>
 DESCARTES_IGNORE_WARNINGS_PUSH
 #include <vector>
+#include <memory>
 #include <Eigen/Geometry>
 DESCARTES_IGNORE_WARNINGS_POP
 
 namespace descartes_light
 {
-/** @brief Definition of a state */
 template <typename FloatType>
-using State = Eigen::Matrix<FloatType, Eigen::Dynamic, 1>;
+using Array = Eigen::Array<FloatType, Eigen::Dynamic, 1>;
 
+/** @brief Definition of a state */
 template <typename FloatType>
-using Array = Eigen::Array<FloatType, Eigen::Dynamic, 1>;
+class State
+{
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  using Ptr = std::shared_ptr<State>;
+  using ConstPtr = std::shared_ptr<const State>;
+
+  State() = default;
+  virtual ~State() = default;
+  State(const State&) = delete;
+  State& operator=(const State&) = delete;
+  State(State&&) noexcept = delete;
+  State& operator=(State&&) noexcept = delete;
+
+  ///////////////////////////
+  // Joint State Container //
+  ///////////////////////////
+  Eigen::Matrix<FloatType, Eigen::Dynamic, 1> values;
```

Comment:
I think if we're going to expose all possible Eigen operations for the class itself, we should hide the actual Eigen object to make it less confusing about which object to do operations with

```suggestion
private:
  Eigen::Matrix<FloatType, Eigen::Dynamic, 1> values;
```

Reply:
I made it protected so derived classes can access the member variable. Any issues?

Reply:
Well now that I think about it. If we make it hidden then we must expose methods of Eigen as needed. For example, if you wanted to `.array()` which is currently not exposed. Is this ok? Or if we leave it public then you have full access to the type if you need something special like calling `.array()`.

Reply:
Let's leave it protected for now. The rest of the `descartes` code seems to build/work with the functions that are currently exposed. We can add `.array` and others later as needed